bundles/org.openhab.binding.homekit/src/main/java/org/openhab/binding/homekit/internal/crypto/CryptoUtils.java

```diff
@@ -0,0 +1,218 @@
+        HKDFBytesGenerator hkdf = new HKDFBytesGenerator(new SHA512Digest());
+        hkdf.init(new HKDFParameters(inputKey, salt, info));
+        byte[] output = new byte[32];
+        hkdf.generateBytes(output, 0, output.length);
+        return output;
+    }
+
+    /**
+     * Generates a 64 bit nonce using the given counter.
+     *
+     * @param counter The counter value.
+     * @return The generated nonce.
+     */
+    public static byte[] generateNonce64(int counter) {
+        return ByteBuffer.allocate(8).order(ByteOrder.LITTLE_ENDIAN).putLong(counter).array();
+    }
+
+    // Compute shared secret using ECDH
+    public static byte[] generateSharedSecret(X25519PrivateKeyParameters clientEphemeralSecretKey,
+            X25519PublicKeyParameters serverEphemeralPublicKey) {
+        byte[] secret = new byte[32];
+        clientEphemeralSecretKey.generateSecret(serverEphemeralPublicKey, secret, 0);
+        return secret;
+    }
+
+    // Generate ephemeral X25519 (Curve25519) key pair
+    public static X25519PrivateKeyParameters generateX25519KeyPair()
+            throws NoSuchAlgorithmException, NoSuchProviderException {
+        return new X25519PrivateKeyParameters(new SecureRandom());
+    }
+
+    public static byte[] sha512(byte[] data) throws NoSuchAlgorithmException {
+        MessageDigest md = MessageDigest.getInstance("SHA-512");
+        return md.digest(data);
+    }
+
+    // Sign message with Ed25519
+    public static byte[] signMessage(Ed25519PrivateKeyParameters secretKey, byte[] message) {
+        Ed25519Signer signer = new Ed25519Signer();
+        signer.init(true, secretKey);
+        signer.update(message, 0, message.length);
+        return signer.generateSignature();
+    }
+
+    public static BigInteger toBigInteger(String hexBlock) throws IllegalArgumentException {
+        String plainHex = hexBlock.replaceAll("\\s+", "");
+        if (plainHex.length() % 2 != 0) {
+            throw new IllegalArgumentException("Hex string must have even length");
+        }
+        return new BigInteger(plainHex, 16);
+    }
+
+    public static byte[] toBytes(String hexBlock) throws IllegalArgumentException {
+        String plainHex = hexBlock.replaceAll("\\s+", "");
+        if (plainHex.length() % 2 != 0) {
+            throw new IllegalArgumentException("Hex string must have even length");
+        }
+        int length = plainHex.length();
+        byte[] result = new byte[length / 2];
+        for (int i = 0; i < length; i += 2) {
+            int hi = Character.digit(plainHex.charAt(i), 16);
+            int lo = Character.digit(plainHex.charAt(i + 1), 16);
+            if (hi == -1 || lo == -1) {
+                throw new IllegalArgumentException(
+                        "Invalid hex character: " + plainHex.charAt(i) + plainHex.charAt(i + 1));
+            }
+            result[i / 2] = (byte) ((hi << 4) + lo);
+        }
+        return result;
+    }
```

Comment:
Can this be used instead, or are they doing something different?

https://github.com/openhab/openhab-core/blob/45cbdbec04ea77ed97934b13e4e2cd2bcbe57aba/bundles/org.openhab.core/src/main/java/org/openhab/core/util/HexUtils.java#L87-L114

Reply:
Do you think it would be worth adding that to `HexUtils` in core?

Reply:
Only today did I notice that actually this is only used by tests, so it could have been moved to the test folder. Now we have it in core though, hopefully others can benefit from it. 👍